api/v1alpha4/clusterclass_webhook.go

```diff
@@ -170,10 +148,38 @@ func (in *ClusterClass) validate(old *ClusterClass) error {
 				),
 			)
 		}
-		oldClassNames.Insert(oldClass.Class)
 	}
 
-	// Makes sure no additional changes were applied.
+	// ensure no previous machine deployment class was modified
+	for _, class := range in.Spec.Workers.MachineDeployments {
+		for _, oldClass := range old.Spec.Workers.MachineDeployments {
+			if class.Class == oldClass.Class && !reflect.DeepEqual(class, oldClass) {
+				allErrs = append(allErrs,
+					field.Invalid(
+						field.NewPath("spec", "workers", "machineDeployments"),
+						class,
+						"cannot be changed.",
+					),
+				)
+			}
+		}
+	}
+
+	return allErrs
+}
+
+func (in ClusterClass) validateCompatibleSpecChanges(old *ClusterClass) field.ErrorList {
```

Comment:
Should we validate ControlPlane changes too?

Reply:
ControlPlane changes are already checked. In the initial implementation we make sure control plane is immutable.
Ref: https://github.com/ykakarap/cluster-api/blob/clusterclass_validations/api/v1alpha4/clusterclass_webhook.go#L193-L201.

Similarly, we make sure that infrastructure is also immutable.

Reply:
Will mark the conversation as resolved after @fabriziopandini confirms that immutability checks are enough for now.

Reply:
sounds good!